src/lightning/data/cache/dataset_optimizer.py

```diff
@@ -0,0 +1,761 @@
+import logging
+import os
+import signal
+import traceback
+from abc import ABC, abstractmethod
+from enum import Enum
+from multiprocessing import Process, Queue
+from pathlib import Path
+from queue import Empty
+from shutil import copyfile
+from threading import Thread
+from time import sleep, time
+from typing import Any, Callable, Dict, List, Literal, Optional, Tuple
+from urllib import parse
+
+from tqdm import tqdm
+
+from lightning import seed_everything
+from lightning.data.cache import Cache
+from lightning.data.cache.constants import (
+    _BOTO3_AVAILABLE,
+    _DEFAULT_FAST_DEV_RUN_ITEMS,
+    _INDEX_FILENAME,
+    _LIGHTNING_CLOUD_GREATER_EQUAL_0_5_41,
+    _TORCH_GREATER_EQUAL_2_1_0,
+)
+
+if _TORCH_GREATER_EQUAL_2_1_0:
+    from torch.utils._pytree import tree_flatten, tree_unflatten
+
+if _LIGHTNING_CLOUD_GREATER_EQUAL_0_5_41:
+    from lightning_cloud.resolver import _LightningSrcResolver, _LightningTargetResolver
+
+if _BOTO3_AVAILABLE:
+    import boto3
+
+logger = logging.Logger(__name__)
+
+
+def _get_cache_folder() -> str:
+    """Returns the cache folder."""
+    return os.getenv("CACHE_FOLDER", "/cache")
+
+
+def _get_num_nodes() -> int:
+    """Returns the number of nodes."""
+    return int(os.getenv("NUM_NODES", 1))
```

Comment:
Note that NUM_NODES is not a standard env var that's generally set. Defaulting it to 1 whatever calls `_get_num_nodes()` might lead to unexpected results.

Reply:
Added `DATA_OPTIMIZER` so folks don't get confused.